graylog2-server/src/main/java/org/graylog2/plugin/inputs/transports/util/KeyUtil.java

```diff
@@ -174,4 +199,76 @@ public static PrivateKey readPrivateKey(Path path) throws IOException {
         final byte[] bytes = Files.readAllBytes(path);
         return PemPrivateKey.valueOf(bytes);
     }
+
+    /**
+     * Build a password-encrypted PKCS8 private key and write it to a PEM file in the temp directory.
+     * Caller is responsible for ensuring that the temp directory is writable. The file will be deleted
+     * when the VM exits.
+     * @param password to protect the key
+     * @param key encrypt this key
+     * @return PEM file
+     * @throws GeneralSecurityException
+     */
+    public static File generatePKSC8PrivateKey(char[] password, PrivateKey key) throws GeneralSecurityException {
+        try {
+            JceOpenSSLPKCS8EncryptorBuilder encryptorBuilder =
+                    new JceOpenSSLPKCS8EncryptorBuilder(PKCS8Generator.AES_256_CBC)
+                            .setRandom(new SecureRandom())
+                            .setPasssword(password);
+            OutputEncryptor encryptor = encryptorBuilder.build();
+
+            // construct object to create the PKCS8 object from the private key and encryptor
+            PemObject pemObj = new JcaPKCS8Generator(key, encryptor).generate();
+            StringWriter stringWriter = new StringWriter();
+            try (JcaPEMWriter pemWriter = new JcaPEMWriter(stringWriter)) {
+                pemWriter.writeObject(pemObj);
+            }
+
+            // write PKCS8 to file
+            String pkcs8Key = stringWriter.toString();
+            File tmpFile = File.createTempFile("key", ".key");
+            tmpFile.deleteOnExit();
+            try (FileOutputStream fos = new FileOutputStream(tmpFile)) {
+                fos.write(pkcs8Key.getBytes(StandardCharsets.UTF_8));
+                fos.flush();
+            }
+            return tmpFile;
+        } catch (IOException | OperatorCreationException e) {
+            throw new GeneralSecurityException(e);
+        }
+    }
+
+    /**
+     * Obtain a private key from a PKS8 PEM file, which is optionally password-protected.
+     * @param password password to decrypt the file - it may be null or empty in case of an unencrypted file
+     * @param keyFile the key file
+     * @return the corresponding private key
+     */
+    public static PrivateKey privateKeyFromFile(String password, File keyFile) throws IOException, PKCSException, OperatorCreationException {
+        PrivateKey privateKey;
+
+        JcaPEMKeyConverter converter = new JcaPEMKeyConverter().setProvider("BC");
+
+        // Be sure to specify charset for reader - don't use plain FileReader
+        Object object;
+        final InputStream inputStream = Files.newInputStream(keyFile.toPath());
```

Comment:
I would put this in a try-with block, because that actually opens a file

Reply:
I've actually meant to wrap the `InputStream`, not the `InputStreamReader` in a try with, since that's the one holding the file handle